src/PhpSpec/ServiceContainer/IndexedServiceContainer.php

```diff
@@ -0,0 +1,238 @@
+     * @param callable $callable
+     * @param array    $tags
+     *
+     * @throws \InvalidArgumentException if service is not a callable
+     */
+    public function define($id, $callable, $tags = [])
+    {
+        if (!is_callable($callable)) {
+            throw new InvalidArgumentException(sprintf(
+                'Service should be callable, "%s" given.',
+                gettype($callable)
+            ));
+        }
+
+        $this->addToIndex(
+            $id,
+            function ($container) use ($callable) {
+                static $instance;
+
+                if (null === $instance) {
+                    $instance = call_user_func($callable, $container);
+                }
+
+                return $instance;
+            },
+            $tags
+        );
+    }
+
+    /**
+     * Retrieves a service from the container
+     *
+     * @param string $id
+     *
+     * @return object
+     *
+     * @throws \InvalidArgumentException if service is not defined
+     */
+    public function get($id)
+    {
+        if (!array_key_exists($id, $this->services)) {
+            throw new InvalidArgumentException(sprintf('Service "%s" is not defined.', $id));
+        }
+
+        return call_user_func($this->services[$id], $this);
+    }
+
+    /**
+     * Determines whether a service is defined
+     *
+     * @param $id
+     * @return bool
+     */
+    public function has($id)
+    {
+        return array_key_exists($id, $this->services);
+    }
+
+    /**
+     * Removes a service from the container
+     *
+     * @param string $id
+     *
+     * @throws \InvalidArgumentException if service is not defined
+     */
+    public function remove($id)
+    {
+        if (!array_key_exists($id, $this->services)) {
+            throw new InvalidArgumentException(sprintf('Service "%s" is not defined.', $id));
+        }
+
+        unset($this->services[$id]);
+    }
+
+    /**
+     * Adds a service or service definition to the index
+     *
+     * @param string $id
+     * @param mixed  $value
+     * @param array  $tags
+     */
+    private function addToIndex($id, $value, array $tags)
+    {
+        foreach ($tags as $tag) {
+            if (!isset($this->tags[$tag])) {
+                $this->tags[$tag] = [];
+            }
+            $this->tags[$tag][] = $id;
+        }
+
+        $this->services[$id] = $value;
+    }
+
+    /**
+     * Finds all services tagged with a particular string
+     *
+     * @param string $tag
+     *
+     * @return array
+     */
+    public function getByTag($tag)
+    {
+        return array_map([$this, 'get'], isset($this->tags[$tag]) ? $this->tags[$tag] : []);
+    }
+
+    /**
+     * Adds a configurator, that can configure many services in one callable
+     *
+     * @internal
+     *
+     * @param callable $configurator
+     *
+     * @throws \InvalidArgumentException if configurator is not a callable
+     */
+    public function addConfigurator($configurator)
```

Comment:
As you started typehinting the definition callable, you could use a typehint here too